pkg/operation/seed/scheduler/gardener_kube_scheduler_test.go

```diff
@@ -0,0 +1,440 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler_test
+
+import (
+	"context"
+
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	. "github.com/gardener/gardener/pkg/operation/seed/scheduler"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+
+	"github.com/Masterminds/semver"
+	resourcesv1alpha1 "github.com/gardener/gardener-resource-manager/pkg/apis/resources/v1alpha1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Scheduler", func() {
+	const (
+		deployNS = "test-namespace"
+		image    = "foo:v1"
+	)
+
+	var (
+		ctx                                              context.Context
+		c                                                client.Client
+		sched                                            component.DeployWaiter
+		version                                          *semver.Version
+		webhookClientConfig, expectedWebhookClientConfig *admissionregistrationv1beta1.WebhookClientConfig
+		expectedLabels                                   map[string]string
+		codec                                            serializer.CodecFactory
+		assertScheduler                                  = func() {
+			Context("New succeeds", func() {
+				var managedResourceSecret *corev1.Secret
+
+				BeforeEach(func() {
+					managedResourceSecret = &corev1.Secret{}
+				})
+
+				JustBeforeEach(func() {
+					s, err := New(c, deployNS, image, version, &dummyConfigurator{}, webhookClientConfig)
+					Expect(err).NotTo(HaveOccurred(), "New succeeds")
+
+					sched = s
+				})
+
+				Context("Deploy", func() {
+					JustBeforeEach(func() {
+						Expect(sched.Deploy(ctx)).NotTo(HaveOccurred(), "deploy succeeds")
+
+						Expect(c.Get(ctx, types.NamespacedName{
+							Name:      "managedresource-gardener-kube-scheduler",
+							Namespace: deployNS,
+						}, managedResourceSecret)).NotTo(HaveOccurred(), "can get managed resource's secret")
+					})
+
+					It("namespace is created", func() {
+						actual := &corev1.Namespace{}
+						expected := &corev1.Namespace{
+							ObjectMeta: metav1.ObjectMeta{
+								Name:   deployNS,
+								Labels: expectedLabels,
+							},
+						}
+						Expect(c.Get(ctx, types.NamespacedName{Name: deployNS}, actual)).NotTo(HaveOccurred(), "can get namespace")
```

Comment:
I like it, but, unfortunately, I don't get it: How does this test work? In `Deploy` you only create the `ManagedResource` with the fake client. Who creates the resources managed by the `ManagedResource` so that you can `Get()` them here with this fake client?

Reply:
I don't get "real" resources just get the resources stored in the secret for that managed resource and then decode them into the correct K8S Object.

Reply:
Ah, I see it now (further below), thanks, makes perfect sense and looks great.

Reply:
The namespace is created manually, because `ManagedResource` cannot be created in not-existing namespace.